cmd/mqtt-log-stdout/main.go

```diff
@@ -47,41 +52,73 @@ func start(cfg config.Client) error {
 	metricsServer := newMetricsServer(cfg, statusClient)
 	mqttClient := newMqttClient(cfg, statusClient, messageClient)
 
-	go metricsServer.Start()
+	start(ctx, g, metricsServer)
+	start(ctx, g, mqttClient)
 
-	var result error
-	err := mqttClient.Start()
-	if err != nil {
-		statusClient.Print("Received error starting mqtt client", err)
-		result = multierror.Append(result, err)
-	}
+	stoppedBy := waitForStop(stopChan, ctx)
+	statusClient.Print(fmt.Sprintf("Application stopping, initiated by: %s", stoppedBy), nil)
 
-	stoppedBy := func() string {
-		select {
-		case sig := <-stopChan:
-			return fmt.Sprintf("os.Signal (%s)", sig)
-		case <-mqttClient.Done():
-			return "mqtt client"
-		case <-metricsServer.Done():
-			return "metrics server"
-		}
-	}()
+	cancel()
 
-	statusClient.Print(fmt.Sprintf("Application stopping, initiated by: %s", stoppedBy), nil)
+	timeoutCtx, timeoutCancel := newTimeoutContext()
+	defer timeoutCancel()
+
+	stop(timeoutCtx, g, mqttClient)
+	stop(timeoutCtx, g, metricsServer)
 
-	err = mqttClient.Stop()
+	return waitForErrGroup(g)
+}
+
+func newErrGroup() (*errgroup.Group, context.Context, context.CancelFunc) {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	g, ctx := errgroup.WithContext(ctx)
+	return g, ctx, cancel
+}
+
+func waitForErrGroup(g *errgroup.Group) error {
+	err := g.Wait()
 	if err != nil {
-		statusClient.Print("Received error stopping mqtt client", err)
-		result = multierror.Append(result, err)
+		return fmt.Errorf("error groups error: %w", err)
 	}
 
-	err = metricsServer.Stop()
-	if err != nil {
-		statusClient.Print("Received error stopping metrics server", err)
-		result = multierror.Append(result, err)
+	return nil
+}
+
+func newTimeoutContext() (context.Context, context.CancelFunc) {
```

Comment:
Yes and yes. Will rename it and lower the timeout to maybe 2 seconds. The important thing with the timeout here is that if the Stop haven't finished within the timeframe it will be forcefully closed.